src/table/src/predicate.rs

```diff
@@ -66,6 +70,192 @@ impl Predicate {
+        }
+    }
+
+    pub fn build(&self) -> TimestampRange {
+        let mut res = TimestampRange::min_to_max();
+        for expr in self.filters {
+            let range = self
+                .extract_time_range_from_expr(expr.df_expr())
+                .unwrap_or_else(TimestampRange::min_to_max);
+            res = res.and(&range);
+        }
+        res
+    }
+
+    /// Extract time range filter from `WHERE`/`IN (...)`/`BETWEEN` clauses.
+    /// Return None if no time range can be found in expr.
+    fn extract_time_range_from_expr(&self, expr: &DfExpr) -> Option<TimestampRange> {
+        match expr {
+            DfExpr::BinaryExpr(BinaryExpr { left, op, right }) => {
+                self.extract_from_binary_expr(left, op, right)
+            }
+            DfExpr::Between(Between {
+                expr,
+                negated,
+                low,
+                high,
+            }) => self.extract_from_between_expr(expr, negated, low, high),
+            DfExpr::InList {
+                expr,
+                list,
+                negated,
+            } => self.extract_from_in_list_expr(expr, *negated, list),
+            _ => None,
+        }
+    }
+
+    fn extract_from_binary_expr(
+        &self,
+        left: &DfExpr,
+        op: &Operator,
+        right: &DfExpr,
+    ) -> Option<TimestampRange> {
+        match op {
+            Operator::Eq => self
+                .get_timestamp_filter(left, right)
+                .map(TimestampRange::single),
+            Operator::Lt => self
+                .get_timestamp_filter(left, right)
+                .map(|ts| TimestampRange::until_end(ts, false)),
+            Operator::LtEq => self
+                .get_timestamp_filter(left, right)
+                .map(|ts| TimestampRange::until_end(ts, true)),
+            Operator::Gt => self
+                .get_timestamp_filter(left, right)
+                .map(TimestampRange::from_start),
+            Operator::GtEq => self
+                .get_timestamp_filter(left, right)
+                .map(TimestampRange::from_start),
+            Operator::And => {
+                // instead of return none when failed to extract time range from left/right, we unwrap the none into
+                // `TimestampRange::min_to_max`.
+                let left = self
+                    .extract_time_range_from_expr(left)
+                    .unwrap_or_else(TimestampRange::min_to_max);
+                let right = self
+                    .extract_time_range_from_expr(right)
+                    .unwrap_or_else(TimestampRange::min_to_max);
+                Some(left.and(&right))
+            }
+            Operator::Or => {
+                let Some(left) = self
+                    .extract_time_range_from_expr(left) else { return None };
+                let Some(right) = self.extract_time_range_from_expr(right) else { return None };
+                Some(left.or(&right))
+            }
+            Operator::NotEq
```

Comment:
I think `a  != ts` can be considered as ` a < ts and a > ts`, so we can extract the time range too.

Reply:
> I think `a != ts` can be considered as ` a < ts and a > ts`, so we can extract the time range too.

This is the caveat of current solution: it does not support multiple time range evaluation. That's why the `or` operation of `TimestampRange` simply calculates `[min(left.start, right.start), max(left.end, right.end))`.

The problem is `a < ts and a > ts` can not be represented in one time range filter, so here simply returns `None`